File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack=[]
        open_b={'(','{','['}
        closed_b={')','}',']'}
        b_dict={'(':')','{':'}','[':']'}
        for char in s:
            if char in open_b:
                stack.append(char)
            elif char in closed_b:
                if not stack:
                    return False
                else:
                    top_stack=stack.pop()
                    if(b_dict[top_stack]!=char):
                        return False
        if stack:
            return False
        else:
            return True

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isOpen(char c){
            return (c=='('||c=='{'||c=='[');
        }
    public boolean isMatching(char a,char b){
            return (a == '(' && b == ')') || (a == '{' && b == '}') || (a == '[' && b == ']');
    }
    public boolean isValid(String s) {
        Stack<Character> stack =new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            if (isOpen(c)) {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                // char top = stack.pop();
                if (!isMatching(stack.peek(), c)) {
                    return false;
                }
                else{
                    stack.pop();
                }
            }
        }
        
        return stack.isEmpty();
    }
}
